way-cooler-bg/src/main.rs

```diff
@@ -0,0 +1,285 @@
+            panic!("Could not open image file");
+        });
+    let mut image = image.to_rgba();
+    let width = image.width();
+    let height = image.height();
+    let stride = width * 4;
+    let size = stride * height;
+    // TODO Split this into its own function
+    {
+        let pixels = image.enumerate_pixels_mut();
+        for (_x, _y, pixel) in pixels {
+            let alpha = pixel[3] as u32;
+            pixel[0] = weird_math(pixel[0], alpha);
+            pixel[1] = weird_math(pixel[1], alpha);
+            pixel[2] = weird_math(pixel[2], alpha);
+
+            let tmp = pixel[2];
+            pixel[2] = pixel[0];
+            pixel[0] = tmp;
+        }
+    }
+    let vec = image.into_vec();
+    let mut tmp = tempfile::NamedTempFile::new().expect("Unable to create a tempfile.");
+    tmp.set_len(size as u64).expect("Could not truncate length of file");
+    tmp.write_all(&*vec).unwrap();
+
+
+    // TODO I want this shit outta here
+    let shm = env.shm.as_ref().map(|o| &o.0).unwrap();
+
+    // Create the surface we are going to write into
+
+    let pool = shm.create_pool(tmp.as_raw_fd(), size as i32);
+    let background_buffer = pool.create_buffer(0, width as i32, height as i32, stride as i32, WlShmFormat::Argb8888);
+    // Tell Way Cooler not to put this in the tree, treat as background
+    // TODO Make this less hacky by actually giving way cooler access to this thing...
+
+    // Attach the buffer to the surface
+    background_surface.attach(Some(&background_buffer), 0, 0);
+    background_surface.damage(0, 0, width as i32, height as i32);
+    Ok(background_buffer)
+}
+
+fn cursor_surface(cursor_path: &str, cursor_surface: &mut WlSurface, env: &WaylandEnv)
+                  -> Result<CursorBuffer, ()> {
+    let shm = env.shm.as_ref().map(|o| &o.0).unwrap();
+
+    let cursor_theme = load_theme(None, 16, shm);
+    /* If the theme has a predefined cursor, just use that */
+    let cursor_buffer: WlBuffer;
+    if let Some(cursor) = cursor_theme.get_cursor("default") {
+        let cursor_frame_buffer = &*cursor.frame_buffer(0).expect("Couldn't get frame_buffer");
+        cursor_surface.attach(Some(cursor_frame_buffer), 0, 0);
+        ::std::mem::forget(cursor_frame_buffer);
+        return Ok(CursorBuffer::Null)
+    } else {
+        let image = open(cursor_path)
+            .unwrap_or_else(|_| {
+                println!("Could not open \"{:?}\"", cursor_path);
+                panic!("Could not open image file");
+            });
+        let mut image = image.to_rgba();
+        let width = image.width();
+        let height = image.height();
+        let stride = width * 4;
+        let size = stride * height;
+        // TODO Split this into its own function
+        {
+            let pixels = image.enumerate_pixels_mut();
+            for (_x, _y, pixel) in pixels {
+                let alpha = pixel[3] as u32;
+                pixel[0] = weird_math(pixel[0], alpha);
```

Comment:
Ah, I was going to split this into it's own method, and I can't remember why exactly I didn't...I think it was weird type issues because this is from Piston' `Image` library, and it has weird types with a bunch of generics that were not happy with me. I'll take another look at it.